feat(shared-data, api): Support for Fixture validations and low level infastructure

CaseyBatten opened the pull request:
# Overview
Creation of deck definitions for the new v4 Deck Fixtures schema. Designed to mimic behavior of labware definitions. Validations added to allow for checking of various requirements/conditions among the set of Fixtures to be expected.

# Risk assessment
Low level risk as this does not interface with existing software at the moment.

CaseyBatten committed:
deck classes and validations using v4 schema

CaseyBatten committed:
Merge branch 'edge' into Fixture-infastructure-and-validations

CaseyBatten committed:
Fixes to lint issues

sfoster1 reviewed (approved):
Looks good! We might want tests in here but the validations are pretty straightforward and we can add tests as we need, could merge as is IMO

CaseyBatten committed:
changes to models supplied

Realistically the whole deck definition base class will be utilized to create the deck "instance", everything else derives from that.

SyntaxColoring commented on shared-data/python/opentrons_shared_data/deck/deck_definitions.py:
```diff
@@ -0,0 +1,206 @@
+    default: Default
+
+
+class DeckDefinition(BaseModel):
```
Let's call this `DeckDefinitionV4` to disambiguate from the other `TypedDict`s that we have floating around.

CaseyBatten committed:
Naming convention update

SyntaxColoring reviewed (approved):
Agreed with Seth—I think this is good to merge now if it helps unblock you, @jbleon95, or @sanni-t. There are some things like tests and organization that we can continue working on in follow-up PRs.

Thank you!
